### PR TITLE
feat: add support for scaling BPM by different ratios

### LINKED ISSUE #9133: add controls to manipulate the bpm via skin/midi controller
Reported by: **[index82](https://launchpad.net/~index82)**
Date: 2018-02-13T17:35:16Z
Status: Confirmed
Importance: Wishlist
Launchpad Issue: [lp1749252](https://bugs.launchpad.net/bugs/1749252)
Tags: bpm, skin, usability

---

like e.g. beats_adjust_faster, beats_adjust_slower ...

Options:
- double BPM
- halve BPM
- 2/3 BPM
- 3/4 BPM
- 4/3 BPM
- 3/2 BPM

### PR DESCRIPTION
**Pull request by mxmilkiib**
This change introduces the ability to scale the BPM of a track by different ratios using ControlPushButton objects. Each button is connected to a lambda function that calls the slotScaleBpm method with the corresponding BPM scale factor.

### CODE REVIEW

**Comment by daschuer** [comment]
Unfortunately clang-format is still complaining. 
I suggest to install pre-commit locally to format every commit automatically. https://github.com/mixxxdj/mixxx/wiki/Using-Git#set-up-automatic-code-checking
It will also install clang-format. 
You may format your whole PR
By `git clang-format upstream/main`
(Make sure not to format the whole code base.)

File: src/engine/controls/bpmcontrol.cpp [hunk 1]
```diff
@@ -93,6 +93,61 @@ BpmControl::BpmControl(const QString& group,
             this,
             &BpmControl::slotTranslateBeatsMove,
             Qt::DirectConnection);
+    
+    m_pBeatsHalve = new ControlPushButton(ConfigKey(group, "beats_set_halve"), false);
```
Comment (daschuer) [inline src/engine/controls/bpmcontrol.cpp:97]: We try to avoid plain new because it is easy to forget the corresponding delete. 
Please use `std::make_unique`

**Review by daschuer — CHANGES_REQUESTED** [review]

File: src/engine/controls/bpmcontrol.cpp (hunk 1 above)
Comment (ronso0) [reply src/engine/controls/bpmcontrol.cpp:97]: Apart from that this PR works as desired :+1:

Comment (mxmilkiib) [reply src/engine/controls/bpmcontrol.cpp:97]: What would the syntax be? Not `    m_pBeatsHalve = std::make_unique<ControlPushButton>(ControlPushButton(ConfigKey(group, "beats_set_halve"), false));
` or `    m_pBeatsHalve = std::make_unique<ControlPushButton>((ConfigKey(group, "beats_set_halve"), false));`..

Comment (ronso0) [reply src/engine/controls/bpmcontrol.cpp:97]: The latter, exactly as I wrote. just without the double ((
Actually, the boolean `bPersists` can also be omitted since it defaults to `false`.
And we switch to use QStringLiteral()
So it would be
`m_pBeatsHalve = std::make_unique<ControlPushButton>(ConfigKey(group, QStringLiteral("beats_set_halve")));` etc.

Comment (mxmilkiib) [reply src/engine/controls/bpmcontrol.cpp:97]: Ah thanks, I don't know why the issue page didn't auto update with your comment. I'll sort this out in a bit

**Comment by mxmilkiib** [comment]
That syntax gives me this though;

```
In file included from /home/milk/src/mixxx/src/engine/controls/bpmcontrol.cpp:1:
/home/milk/src/mixxx/src/engine/controls/bpmcontrol.h:149:39: error: template argument 1 is invalid
  149 |     std::unique_ptr<ControlPushButton m_pBeatsHalve;
      |                                       ^~~~~~~~~~~~~
/home/milk/src/mixxx/src/engine/controls/bpmcontrol.h:149:39: error: template argument 2 is invalid
/home/milk/src/mixxx/src/engine/controls/bpmcontrol.h:149:10: error: ‘<expression error>’ in namespace ‘std’ does not name a type
  149 |     std::unique_ptr<ControlPushButton m_pBeatsHalve;
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
...
/home/milk/src/mixxx/src/engine/controls/bpmcontrol.cpp: In constructor ‘BpmControl::BpmControl(const QString&, UserSettingsPointer)’:
/home/milk/src/mixxx/src/engine/controls/bpmcontrol.cpp:97:5: error: ‘m_pBeatsHalve’ was not declared in this scope; did you mean ‘m_pBeats’?
   97 |     m_pBeatsHalve = std::make_unique<ControlPushButton>(ConfigKey(group, "beats_set_halve"), false);
      |     ^~~~~~~~~~~~~
      |     m_pBeats
```

**Comment by ronso0** [comment]
> std::unique_ptr<ControlPushButton m_pBeatsHalve;

missing closing `>`

**Comment by mxmilkiib** [comment]
with closing bracket;

```
/home/milk/src/mixxx/src/engine/controls/bpmcontrol.cpp: In constructor ‘BpmControl::BpmControl(const QString&, UserSettingsPointer)’:
/home/milk/src/mixxx/src/engine/controls/bpmcontrol.cpp:98:12: error: no matching function for call to ‘BpmControl::connect(std::unique_ptr<ControlPushButton>&, void (ControlObject::*)(double), BpmControl*, BpmControl::BpmControl(const QString&, UserSettingsPointer)::<lambda(int)>)’
   98 |     connect(m_pBeatsHalve,
      |     ~~~~~~~^~~~~~~~~~~~~~~
   99 |             &ControlObject::valueChanged,
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  100 |             this,
      |             ~~~~~
  101 |             [this](int value) {
      |             ~~~~~~~~~~~~~~~~~~~
  102 |                 if (value > 0) {
      |                 ~~~~~~~~~~~~~~~~
  103 |                     slotScaleBpm(mixxx::Beats::BpmScale::Halve);
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  104 |                 }
      |                 ~
  105 |             });
      |             ~~
In file included from /usr/include/qt6/QtCore/QObject:1,
                 from /home/milk/src/mixxx/src/control/controlobject.h:3,
                 from /home/milk/src/mixxx/src/engine/controls/bpmcontrol.h:5,
                 from /home/milk/src/mixxx/src/engine/controls/bpmcontrol.cpp:1:
/usr/include/qt6/QtCore/qobject.h:198:9: note: candidate: ‘static QMetaObject::Connection QObject::connect(const typename QtPrivate::FunctionPointer<Prototype>::Object*, Func1, const typename QtPrivate::ContextTypeForFunctor<Func2>::ContextType*, Func2&&, Qt::ConnectionType) [with Func1 = void (ControlObject::*)(double); Func2 = BpmControl::BpmControl(const QString&, UserSettingsPointer)::<lambda(int)>; typename QtPrivate::FunctionPointer<Prototype>::Object = ControlObject; typename QtPrivate::ContextTypeForFunctor<Func2>::ContextType = QObject]’
  198 |         connect(const typename QtPrivate::FunctionPointer<Func1>::Object *sender, Func1 signal,
      |         ^~~~~~~
/usr/include/qt6/QtCore/qobject.h:198:75: note:   no known conversion for argument 1 from ‘std::unique_ptr<ControlPushButton>’ to ‘const QtPrivate::FunctionPointer<void (ControlObject::*)(double)>::Object*’ {aka ‘const ControlObject*’}
  198 |         connect(const typename QtPrivate::FunctionPointer<Func1>::Object *sender, Func1 signal,
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~
/usr/include/qt6/QtCore/qobject.h:246:9: note: candidate: ‘template<class Func1, class Func2> static QMetaObject::Connection QObject::connect(const typename QtPrivate::FunctionPointer<Prototype>::Object*, Func1, Func2&&)’
  246 |         connect(const typename QtPrivate::FunctionPointer<Func1>::Object *sender, Func1 signal, Func2 &&slot)
      |         ^~~~~~~
/usr/include/qt6/QtCore/qobject.h:246:9: note:   template argument deduction/substitution failed:
/home/milk/src/mixxx/src/engine/controls/bpmcontrol.cpp:98:12: note:   candidate expects 3 arguments, 4 provided
   98 |     connect(m_pBeatsHalve,
      |     ~~~~~~~^~~~~~~~~~~~~~~
   99 |             &ControlObject::valueChanged,
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  100 |             this,
      |             ~~~~~
  101 |             [this](int value) {
      |             ~~~~~~~~~~~~~~~~~~~
  102 |                 if (value > 0) {
      |                 ~~~~~~~~~~~~~~~~
  103 |                     slotScaleBpm(mixxx::Beats::BpmScale::Halve);
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  104 |                 }
      |                 ~
  105 |             });
      |             ~~
/usr/include/qt6/QtCore/qobject.h:177:36: note: candidate: ‘static QMetaObject::Connection QObject::connect(const QObject*, const char*, const QObject*, const char*, Qt::ConnectionType)’
  177 |     static QMetaObject::Connection connect(const QObject *sender, const char *signal,
      |                                    ^~~~~~~
...
```

**Comment by mxmilkiib** [comment]
Aah cmake cache again..

**Comment by ronso0** [comment]
Yes, it's a unique_ptr now so we need to `get()` the value first (the ControlPushButton in this case).
This should work:
```
connect(m_pBeatsHalve.get(),
       &ControlObject::valueChanged,
       this,
       [this](int value) {
           if (value > 0) {
               slotScaleBpm(mixxx::Beats::BpmScale::Halve);
           }
       });
```

**Comment by mxmilkiib** [comment]
The style changes mentioned are inconsistent, I could do the extra line breaks, but it's not asking for them in all cases..?

**Comment by ronso0** [comment]
Builds are currently failing because the explicit `delete` calls are still in the destructor `BpmControl::~BpmControl` but not applicable to the now unique_ptr.
Please remove those and amend to the respective commits (each commit must build in order to allow git bisect).

In case you still can't build locally I can provide fixups and push to your branch.

**Comment by mxmilkiib** [comment]
Can do, will be afk for a couple of days tho

**Comment by ronso0** [comment]
_ping_ clang-format wants to break some long lines.
> Unfortunately clang-format is still complaining. I suggest to install pre-commit locally to format every commit automatically. https://github.com/mixxxdj/mixxx/wiki/Using-Git#set-up-automatic-code-checking It will also install clang-format. You may format your whole PR By `git clang-format upstream/main` (Make sure not to format the whole code base.)

File: src/engine/controls/bpmcontrol.cpp [hunk 2]
```diff
@@ -1,4 +1,4 @@
-#include "engine/controls/bpmcontrol.h"
+#include "bpmcontrol.h"
```
Comment (ronso0) [inline src/engine/controls/bpmcontrol.cpp:1]: Why did you change this?
We use the full path for all other includes.

**Comment by mxmilkiib** [comment]
> _ping_ clang-format wants to break some long lines.

"The style changes mentioned are inconsistent, I could do the extra line breaks, but it's not asking for them in all cases..?"

Unless I misread then, and I havn't relooked yet - I'll do so after I've done the other thing - it was wanting me to remove line breaks, and for some reason only wanting me to add line breaks to some of the code I'd added thus I assumed that adding further linebreaks would hit me with a requirement to remove those new line breaks too. Maybe I was wrong. I'll suffer trying to fix it multiple times again and it continually giving me inconsistent output until I rage quit and go to bed.

Also, running locally give no problematic output. I guess either I'm doing something wrong or the instructions are inadequate.

image

Well, it is giving a silly number of "skipped", obviously part of the problem, given it should be giving me an error, but I'm not sure how I've fluffed the process.

**Comment by ronso0** [comment]
It's skipping because you have no file changes for pre-commit to check.
clang-format has a line length threshold above which it proposes line-breaks (which may sometimes look rather strange to humans).

You can do
`git log -1` (just to note the commit message)
`git reset HEAD^1`
`git add --all`
`git commit`
Then pre-commit should run again on the changeset.

(by the way I very much recommend the zsh `git` plugin, it has many useful shortkeys)

**Comment by ronso0** [comment]
now it wants to move the header include...

please revert and re-commit as described above, actually pre-commit should only act on changed lines.
then force-push.

**Comment by ronso0** [comment]
...and trailing whitespaces :laughing: 
You can also download the pre-commit patch from the CI summary (at the bottm of the Summary page https://github.com/mixxxdj/mixxx/actions/runs/8242566098?pr=12934), apply that (git commit --amend) and push again.

**Comment by mxmilkiib** [comment]
The error just flip-flopped on line 74 of bpmcontrol.cpp, telling me to unbreak things, now telling me to break things again. I'm going to go and do something else....

**Comment by ronso0** [comment]
> The error just flip-flopped on line 74 of bpmcontrol.cpp, telling me to unbreak things, now telling me to break things again. 

Well, sometimes it's strange.
Did you apply the patch? (or did as the patch suggested)

Personally I'd go with the 4-/5-line version, if it then still complains we'll ignore it.

**Comment by ronso0** [comment]
Re https://github.com/mixxxdj/mixxx/pull/12934/commits/85e97bacb4df4e20b9a6d88e29f2f00e68daae91
You can simply merge upstream/main (mixxxdj/mixxx:main), no need to create PRs for your branch.

**Comment by mxmilkiib** [comment]
Right o. Too computered out now today, shall figure that process properly in the next day or few! Cheers

**Comment by mxmilkiib** [comment]
I guess I should undo the update PR and redo the commit then force push?

**Comment by ronso0** [comment]
I'm the only reviewer until now and I'm fine with cherry-picking / rebasing and amend the commit of #12941.

**Comment by mxmilkiib** [comment]
Cool cool.

I was using Git Butler to "undo" and force push the PR commits, now it has [stopped working](https://github.com/gitbutlerapp/gitbutler/issues/3309), as I made a mess trying to get multiple feature PRs into the same branch so removed the repo dir and recloned but apparently broke some part of a workflow. I learnt some `git` years ago but have forgotten most, and never rebased nor force pushed back then, let alone cherry picked commits, so am rather lost in how to explain to myself what I need to do. 

image

I also retrieved both my PRs with `gh pr checkout` then tried `git cherry [sha of main commit in first PR]`, but that just returned another SHA looking string, and I can't tell if anything changed.

**Comment by mxmilkiib** [comment]
Ah, `git cherry-pick` not `git cherry`.

Lets see how wrong I get this..

**Comment by mxmilkiib** [comment]
A strange sort order;
image

Needs a new issue?

**Comment by ronso0** [comment]
I assume this is the MIDI Input table. Actions are sorted by control, and IIRC that is
D ouble
F our thirds
H alf
T hree fourths
T wo thirds etc

Yeah please file a new issue!

**Comment by mxmilkiib** [comment]
DEBUG ASSERT
```
...
warning [Controller] ControlObject accessed via deprecated key "[Channel1]" "hotcue_2_enabled" - use "[Channel1]" "hotcue_2_status" instead
warning [Controller] ControlObject accessed via deprecated key "[Channel1]" "hotcue_2_enabled" - use "[Channel1]" "hotcue_2_status" instead
warning [Controller] ControlObject accessed via deprecated key "[Channel1]" "hotcue_2_enabled" - use "[Channel1]" "hotcue_2_status" instead
warning [Controller] ControlObject accessed via deprecated key "[Channel1]" "hotcue_2_enabled" - use "[Channel1]" "hotcue_2_status" instead
warning [Main]      $ connectMenuBar
warning [0x7c0130000b70] Engine thread not scheduled with the real-time policy SCHED_FIFO
warning [Main] $ WindowStateChange: QFlags<Qt::WindowState>(WindowNoState)
warning [Main] $ WindowStateChange IGNORE
warning [Main] $ WindowStateChange: QFlags<Qt::WindowState>(WindowMinimized)
warning [Main] $ WindowStateChange: QFlags<Qt::WindowState>(WindowMinimized)
warning [Main] $ WindowStateChange IGNORE
warning [Main] $ WindowStateChange: QFlags<Qt::WindowState>(WindowNoState)
warning [Main] $ WindowStateChange: QFlags<Qt::WindowState>(WindowNoState)
warning [Main] $ WindowStateChange IGNORE
warning [Main] $ WindowStateChange: QFlags<Qt::WindowState>(WindowNoState)
warning [Main] $ WindowStateChange IGNORE
warning [Main] $ WindowStateChange: QFlags<Qt::WindowState>(WindowNoState)
warning [Main] $ WindowStateChange IGNORE
warning [Main] $ WindowStateChange: QFlags<Qt::WindowState>(WindowNoState)
warning [Main] $ WindowStateChange IGNORE
warning [Main] $ WindowStateChange: QFlags<Qt::WindowState>(WindowNoState)
warning [Main] $ WindowStateChange IGNORE
warning [Main] $ WindowStateChange: QFlags<Qt::WindowState>(WindowNoState)
warning [Main] $ WindowStateChange IGNORE
warning [Main] $ WindowStateChange: QFlags<Qt::WindowState>(WindowNoState)
warning [Main] $ WindowStateChange IGNORE
warning [Controller] ControlDoublePrivate::getControl returning NULL for ( "[Channel2]" , "beats_set_twothirds" )
critical [Controller] DEBUG ASSERT: "flags.testFlag(ControlFlag::NoAssertIfMissing)" in function static QSharedPointer<ControlDoublePrivate> ControlDoublePrivate::getControl(const ConfigKey&, ControlFlags, ControlObject*, bool, bool, bool, double) at ./src/control/control.cpp:203
warning [Controller] ControlDoublePrivate::getControl returning NULL for ( "[Channel2]" , "beats_set_double" )
critical [Controller] DEBUG ASSERT: "flags.testFlag(ControlFlag::NoAssertIfMissing)" in function static QSharedPointer<ControlDoublePrivate> ControlDoublePrivate::getControl(const ConfigKey&, ControlFlags, ControlObject*, bool, bool, bool, double) at ./src/control/control.cpp:203
warning [Controller] ControlDoublePrivate::getControl returning NULL for ( "[Channel2]" , "beats_set_twothirds" )
critical [Controller] DEBUG ASSERT: "flags.testFlag(ControlFlag::NoAssertIfMissing)" in function static QSharedPointer<ControlDoublePrivate> ControlDoublePrivate::getControl(const ConfigKey&, ControlFlags, ControlObject*, bool, bool, bool, double) at ./src/control/control.cpp:203
warning [Controller] ControlDoublePrivate::getControl returning NULL for ( "[Channel2]" , "beats_set_double" )
critical [Controller] DEBUG ASSERT: "flags.testFlag(ControlFlag::NoAssertIfMissing)" in function static QSharedPointer<ControlDoublePrivate> ControlDoublePrivate::getControl(const ConfigKey&, ControlFlags, ControlObject*, bool, bool, bool, double) at ./src/control/control.cpp:203
warning [Controller] ControlDoublePrivate::getControl returning NULL for ( "[Channel2]" , "beats_set_threehalves" )
critical [Controller] DEBUG ASSERT: "flags.testFlag(ControlFlag::NoAssertIfMissing)" in function static QSharedPointer<ControlDoublePrivate> ControlDoublePrivate::getControl(const ConfigKey&, ControlFlags, ControlObject*, bool, bool, bool, double) at ./src/control/control.cpp:203
warning [Controller] ControlDoublePrivate::getControl returning NULL for ( "[Channel2]" , "beats_set_threehalves" )
critical [Controller] DEBUG ASSERT: "flags.testFlag(ControlFlag::NoAssertIfMissing)" in function static QSharedPointer<ControlDoublePrivate> ControlDoublePrivate::getControl(const ConfigKey&, ControlFlags, ControlObject*, bool, bool, bool, double) at ./src/control/control.cpp:203
warning [Main]      .
warning [Main]      setBeatsWhileLocked  old beats: - | new beats: X
warning [Main]        (currently null beats, don't rec)
warning [Main]      .
warning [Main]      .
warning [Main]      setBeatsWhileLocked  old beats: - | new beats: X
warning [Main]        (currently null beats, don't rec)
warning [Main]      .
warning [Main] $ WindowStateChange: QFlags<Qt::WindowState>(WindowNoState)
warning [Main] $ WindowStateChange IGNORE
warning [Main] $ WindowStateChange: QFlags<Qt::WindowState>(WindowNoState)
warning [Main] $ WindowStateChange IGNORE
warning [Main] $ WindowStateChange: QFlags<Qt::WindowState>(WindowNoState)
warning [Main] $ WindowStateChange IGNORE
warning [Main] $ WindowStateChange: QFlags<Qt::WindowState>(WindowNoState)
warning [Main] $ WindowStateChange IGNORE
warning [Main] $ WindowStateChange: QFlags<Qt::WindowState>(WindowNoState)
warning [Main] $ WindowStateChange IGNORE
warning [LibraryScanner 1] QSqlDatabasePrivate::removeDatabase: connection 'MIXXX-2' is still in use, all queries will cease to work.
./mixxx  567.66s user 326.45s system 9% cpu 2:31:08.48 total
```

File: src/engine/controls/bpmcontrol.cpp [hunk 3]
```diff
@@ -264,6 +334,22 @@ void BpmControl::slotTapFilter(double averageLength, int numSamples) {
...
+    }
+    const mixxx::BeatsPointer pBeats = pTrack->getBeats();
+    if (!pBeats) {
+        return;
+    }
+    const auto newBeats = pBeats->tryScale(bpmScale);
+    if (!newBeats) {
+        return;
+    }
+    pTrack->trySetBeats(*newBeats);
```
Comment (ronso0) [inline src/engine/controls/bpmcontrol.cpp:350]:
```suggestion
    pTrack->trySetBeats(newBeats);
```
`newBeats` as a pointer already.

**Review by ronso0 — CHANGES_REQUESTED** [review]
LGTM, except one tiny yet significant character

**Comment by mxmilkiib** [comment]
Please take this off my hands, I don't have the capacity to focus on fixing and finishing this PR. Thanks.

**Comment by ronso0** [comment]
I have just rebased, fixed and force-pushed to your branch.
Hopefully also clang-tidy is happy now.

**Review by ronso0 — APPROVED** [review]

**Comment by ronso0** [comment]
> I have just rebased, fixed and force-pushed to your branch. Hopefully also clang-tidy is happy now.

Since I worked on this, too, someone else needs to review and merge.

**Comment by ronso0** [comment]
I've split the commit into the std::unique_ptr changes and the actual feature commit.

**Comment by daschuer** [comment]
We have a conflict unfortunately

**Review by daschuer — APPROVED** [review]

**Comment by ronso0** [comment]
> We have a conflict unfortunately

resolved. There'll be _some more_ #12104 when this is merged :sweat_drops:

**Comment by daschuer** [comment]
Thank you.

**Comment by ronso0** [comment]
Thanks @mxmilkiib for starting this, and their patience.
